text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,206 @@
+---
+title: Mixed Shadow Mode
+status: DRAFTED
+created_at: 2021-05-07
+updated_at: YYYY-MM-DD
+pr: https://github.com/salesforce/lwc-rfcs/pull/49
+---
+
+# Mixed Shadow Mode
+
+## Introduction
+
+LWC offers a collection of polyfills that implement Shadow DOM features. These polyfills are
+collectively referred to as "synthetic Shadow DOM" and are published to npm under the
+`@lwc/synthetic-shadow` package. Synthetic shadow DOM provides a consistent experience for LWC web
+components across the Salesforce-supported browser matrix.
+
+Native Shadow DOM support has greatly increased since the start of the LWC project but applications
+continue to rely on the synthetic shadow polyfills to maintain backwards-compatibility. The removal
+of these polyfills can result in broken experiences due to dependencies on certain "escape hatches"
+that are not possible in a native Shadow DOM environment (e.g., instrumentation, styling, etc, which
+are discussed below).
+
+Applications that wish to use LWC components in a native shadow context simply need to omit the
+inclusion of the `@lwc/synthetic-shadow` polyfill; however, this is not a realistic option for
+large applications that cannot afford a rewrite.
+
+This proposal opens up an incremental migration path for applications moving to native Shadow DOM by
+allowing the usage of both native and synthetic Shadow DOM in the same document.
+
+## Detailed design
+
+Components will opt-in to this new feature and the default mode will be synthetic mode. From the
+component's perspective, there are two choices of shadow semantics: synthetic mode or native
+mode. The concept of a mixed mode should be transparent and the framework should guarantee this
+through integration testing, which is further discussed below.
+
+A component can signal that it prefers native Shadow DOM by setting a static property called
+`preferNativeShadow` to `true`.
+
+```js
+export default class extends LightningElement {
+    static preferNativeShadow = true;
+}
+```
+
+If this property is not defined, a default value of `false` will be used. The value of this property
+will be read once during component construction and cached for the lifetime of the component. The
+LWC engine will throw an exception if the value of `preferNativeShadow` is not a boolean value.
+
+All polyfills applied by `@lwc/synthetic-shadow` will look to this property as a signal to invoke
+either the polyfilled API or the native browser API. If the browser does not support Shadow DOM
+(e.g., IE11), then LWC will fallback to synthetic mode regardless of which mode the component
+prefers.
+
+Due to observable differences between the two modes, it is likely that components that prefer native

```

Comment:
what will be such hint?

Reply:
How about making it available as a boolean value on `this.template.native`? This would resolve to the falsy value `undefined` for existing components, which works well in terms of backcompat.

Reply:
well, my initial reaction to this sort of feature is that when exposed, people will create a forking logic depending on the mode in this case, is that something we want? I will like to understand better, maybe with a realistic example, why people will need to know about the mode, for what?

Reply:
The idea is that they might need to fork their logic due to the observable differences listed below. I agree that this is something we don't want to expose unless we have to so it doesn't need to go in the first iteration. Should I just remove this mention for now?